Formulate elliptic systems in terms of fluxes and sources

nilsvu opened the pull request:
## Proposed changes

Refactors the elliptic code to a formulation that resembles closely the
flux-conservative form of evolution systems. The system just computes
the fluxes and sources and no longer has to implement the derivative
operator or the normal-dot-fluxes. Instead, they use code shared
between systems, so adding the Elasticity and XCTS systems (next PRs)
is straight-forward.

**Notes to reviewers:**
- This PR changes a lot of lines, but most commits should be straight-forward to review since they do simple renames.
- I tried to split this PR into many small commits, so it's best to step through the commit history when reviewing.
- <s>I could defer the last 4 commits to a follow-up PR if the reviewers prefer. They should be quite easy to review, so it's probably more effective to keep them here. They actually add some important tests for the code in this PR.</s> Deferred this as requested.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu committed:
Add a prefix tag for the FixedSource

This describes the part of the source that is independent
of the dynamic variables

nilsvu committed:
Use FixedSource prefix in elliptic code

nilsvu committed:
Rename impose_boundary_conditions_on_fields to primal_fields

This alias will see more use than just for boundary conditions. It
specifies which of the first-order variables are "primal" and which
are "auxiliary". The latter won't be sourced, for instance.

nilsvu committed:
Only retrieve sources for primal fields in elliptic initialization

Auxiliary fields are never sourced.

nilsvu committed:
Remove unused functions from elliptic solutions

nilsvu committed:
Embed Poisson tags in namespace

nilsvu committed:
Fix some namespaces in elliptic code

nilsvu committed:
Add compute tag for elliptic first-order fluxes

Uses the system equations to generically compute first-order
fluxes.

nilsvu committed:
Use Poisson field gradient for auxiliary variable

It was a mistake to use a vector quantity for the auxiliary variable before. It should be
a covector, which the principal symbol raises to a vector.
We don't need a special tag for this anymore, since we can just use
the gradient prefix.
This change also allows adding fluxes for the Poisson equation on a
non-Euclidean background metric.

nilsdeppe commented:
I would recommend splitting this into 3 or 4 PRs if possible. Ideally PRs should be 400 lines or less. I realize that's sometimes not possible, but considering you have 13 commits it seems like it should be possible to at least downsize this a bit.

nilsvu commented:
@nilsdeppe Ok, I dropped the last 4 commits. Now it's a bunch of simple preparation-commits leading up to "Formulate elliptic system in terms of fluxes and sources". I think I have now moved everything I could out of that commit.

nilsvu committed:
Add first-order elliptic operator

nilsvu committed:
Formulate elliptic system in terms of fluxes and sources

Refactors the elliptic code to a formulation that resembles closely the
flux-conservative form of evolution systems. The system just computes
the fluxes and sources and no longer has to implement the derivative
operator or the normal-dot-fluxes. Instead, they use code shared
between systems.

kidder commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -0,0 +1,71 @@
+                        operator_applied_to_vars,
+                    const db::item_type<div_fluxes_tag>& div_fluxes,
+                    const db::item_type<sources_tag>& sources) noexcept {
+    *operator_applied_to_vars = -1. * div_fluxes + sources;
```
`sources - div_fluxes`

kidder commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -33,52 +33,78 @@ namespace Poisson {
+ * introduce auxiliary degrees of freedom that can be avoided in the
+ * second-order (or _primal_) formulation. Furthermore, the linear operator that
+ * represents the DG discretization for this system is not symmetric and has
+ * both positive and negative eigenvalues (CHECK THIS). These properties further
```
When are you going to `CHECK_THIS` ?

kidder commented:
@geoffrey4444 please review this

nilsvu commented:
@kidder fixed the issues you commented on and accidentally squashed those in, sorry 😅

geoffrey4444 commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -96,10 +96,14 @@ struct InitializeSystem {
-    auto sources = variables_from_tagged_tuple(
+    // We need only retrieve sources for the primal fields, since the auxiliary
+    // fields will never be sourced.
+    db::item_type<sources_tag> sources{num_grid_points, 0.};
```
Safer to initialize to signaling nan?

nilsvu replied:
This initializes the auxiliary field sources to `0.`

geoffrey4444 commented on src/Elliptic/Actions/InitializeSystem.hpp:
```diff
@@ -96,10 +96,14 @@ struct InitializeSystem {
             .variables(inertial_coords,
-                       db::get_variables_tags_list<sources_tag>{}));
+                       db::wrap_tags_in<::Tags::Source,
+                                        typename system::primal_fields>{}));
```
Just to help me understand, where do you actually specify what the source term is (e.g., the charge density for Poisson)?

nilsvu replied:
That's what this bit of code is about. The "fixed sources" (i.e. the `f(x)` in `-Laplace(u(x))=f(x)`) are being retrieved from the analytic solution. You can find examples in `Poisson::Solutions`.

geoffrey4444 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Xcts/Test_ConstantDensityStar.cpp:
```diff
@@ -29,9 +29,7 @@ using auxiliary_field_tags = tmpl::list<
-using source_tags =
-    db::wrap_tags_in<Tags::Source,
-                     tmpl::append<field_tags, auxiliary_field_tags>>;
+using source_tags = db::wrap_tags_in<Tags::Source, field_tags>;
```
Why not FixedSource?

nilsvu replied:
This indeed changes to `FixedSource` once the tag is added in one of the following commits.

geoffrey4444 commented on src/Elliptic/FirstOrderComputeTags.hpp:
```diff
@@ -0,0 +1,100 @@
+          typename PrimalVars = typename System::primal_variables,
+          typename AuxiliaryVars = typename System::auxiliary_variables,
+          typename FluxesArgs = typename FluxesComputer::argument_tags>
+struct FirstOrderFluxesCompute;
```
To help me learn, what do lines 18-24 do here?

nilsvu replied:
- I need to expand the tags typelists into parameter packs, which is why this forward declaration takes typelists, and the template specialization below extracts the tags they contain as a parameter pack.
- I use these defaulted template parameters to retrieve the typelists from the `System` (and they can also be overridden if other tags should be used).

geoffrey4444 commented on tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp:
```diff
@@ -0,0 +1,127 @@
+  static void apply(const gsl::not_null<Scalar<DataVector>*> source_for_field,
+                    const double& an_argument,
+                    const Scalar<DataVector>& field) {
+    get(*source_for_field) = get(field) * an_argument;
```
Maybe do something different for the sources than the tags, like multiply by sqr(an_argument), to make the test more generic?

nilsvu replied:
good idea 👍

geoffrey4444 commented on tests/Unit/Elliptic/Test_FirstOrderComputeTags.cpp:
```diff
@@ -0,0 +1,127 @@
+              .get(d) == DataVector{num_points, 3. * (d + 1.)});
+    CHECK(get<::Tags::Flux<AuxiliaryFieldTag<Dim>, tmpl::size_t<Dim>,
+                           Frame::Inertial>>(box)
+              .get(d, d) == DataVector{num_points, 6.});
```
Maybe it's too trivial to be worth it, but maybe check that the terms you didn't set are actually zero?

nilsvu replied:
good point, added a check

geoffrey4444 reviewed (changes requested):
Started reviewing. Looked at commits 85e0312 through 2b4cb75. Just some minor comments and some questions to help me learn

geoffrey4444 commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -0,0 +1,71 @@
+ * \ingroup ActionsGroup
+ * \ingroup DiscontinuousGalerkinGroup
+ * \brief Compute the bulk contribution to the operator represented by the
+ * `StepPrefixTag` applied to the `VarsTag`.
```
What is a StepPrefixTag? What possible values can it take? Please specify here or point to relevant docs

nilsvu replied:
Added more details to the docs

kidder replied:
I still don't understand what a StepPrefixTag is.  Can you point to an equation illustrating it?

nilsvu replied:
pushed a fixup to address this, please take a look

geoffrey4444 commented on tests/Unit/Elliptic/Test_FirstOrderOperator.cpp:
```diff
@@ -0,0 +1,93 @@
+
+struct SomeField : db::SimpleTag {
+  using type = Scalar<DataVector>;
+  static std::string name() noexcept { return "SomeField"; }
```
Do you have to specify the `name()` if it's the same as the struct's name?

nilsvu replied:
DataBox tag names don't default to the class names (yet).

wthrowe replied:
Yes they do.  That feature was added with PDALs (#1535).  (Some things still try to call `Tag::name()` directly instead of `db::tag_name<Tag>()`, but those are easy-to-fix bugs.)

nilsvu replied:
You're right of course. I didn't realize `db::tag_name` had that functionality 👍

geoffrey4444 commented on tests/Unit/Elliptic/Test_FirstOrderOperator.cpp:
```diff
@@ -0,0 +1,93 @@
+template <size_t Dim>
+struct AnotherField : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim>;
+  static std::string name() noexcept { return "SomeField"; }
```
Shouldn't this name be `AnotherField`?

geoffrey4444 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/Poisson/Moustache.py:
```diff
@@ -7,25 +7,28 @@
             x[d - 1] * (1. - x[d - 1]) * (poly(x[d], [
                 0.5, poly(x[d - 1], [-3.5, 2, -2]), 7.5, -5
-                ]) + poly(x[d - 1], [0, -1, 1]))
+            ]) + poly(x[d - 1], [0, -1, 1]))
```
Just a nitpick: I'm unclear why the indentation changed here

nilsvu replied:
`autopep8` probably did that. I can revert the change

nilsvu replied:
actually, since `autopep8` wants this formatting, I'll just keep it like this if that's ok

geoffrey4444 commented on src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp:
```diff
@@ -30,45 +34,50 @@ namespace Actions {
+      // data that is being set there to impose homogeneous Dirichlet boundary
+      // conditions. Then, we compute their normal-dot-fluxes. The flux
+      // divergences are sliced from the volume.
+      ::Tags::InterfaceComputeItem<
```
Can you please briefly specify what will need to be extended to implement other BCs than homogeneous Dirichlet?

nilsvu replied:
This is actually valid for any boundary conditions, adjusted the comment. It's the actions in Elliptic/DG that need to be changed to support other boundary conditions (they are not affected by this PR). They can currently handle inhomogeneous Dirichlet b.c.

geoffrey4444 commented on src/Elliptic/Systems/Poisson/Equations.cpp:
```diff
@@ -23,59 +23,31 @@
-  // The minus sign is to cancel the one in `lift_flux`
+void auxiliary_fluxes(gsl::not_null<tnsr::IJ<DataVector, Dim, Frame::Inertial>*>
+                          flux_for_gradient,
+                      const Scalar<DataVector>& field) noexcept {
```
How would you handle an elliptic equation whose fields include a vector field? I'm just a little confused why this is called `field` instead of `scalar_field`

nilsvu replied:
This is the implementation for the Poisson system, so I'm working with the (scalar) `Poisson::Tags::Field` and its gradient. An example for an elliptic system that uses a vector field is the Elasticity system that I'll add once these preparatory PRs are in. It implements its own fluxes and sources for the `Elasticity::Tags::Displacement` vector field and the `Elasticity::Tags::Strain` symmetric rank-2 tensor.

geoffrey4444 commented on src/Elliptic/Systems/Poisson/Equations.cpp:
```diff
@@ -23,59 +23,31 @@
     const gsl::not_null<Variables<package_tags>*> packaged_data,
     const Scalar<DataVector>& field,
-    const tnsr::i<DataVector, Dim, Frame::Inertial>& grad_field,
+    const tnsr::I<DataVector, Dim, Frame::Inertial>& grad_field,
```
Can you please explain why this change? I thought in a previous commit in this PR you changed the gradient to a lower index. In terms of math, I would have thought the gradient of a field would be a convector

nilsvu replied:
This quantity is now the principal symbol applied to the gradient, i.e. for the Euclidean Poisson system it's just the gradient with its index raised with the flat metric. I didn't change the variable name because for the Poisson system it's still the same values and I want to keep changes to the flux to a minimum. That's because I'll delete this Poisson-specific implementation of the flux in #1671 and replace it with a generic internal penalty flux for elliptic systems.

geoffrey4444 commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -33,52 +33,77 @@ namespace Poisson {
  *
- * \details This system formulates the Poisson equation \f$-\Delta u(x) =
- * f(x)\f$ as the set of coupled first-order PDEs
+ * \details This system formulates the Poisson equation \f$-\Delta_\gamma u(x) =
```
Just a nitpick, but I prefer \nabla^2 to \Delta for the Laplacian operator

nilsvu replied:
It's `\Delta` in a lot of places in the documentation already and I don't see a strong reason to change that, in particular because also much of the (NR) literature uses `\Delta`. It's a matter of personal preference I guess.

geoffrey4444 replied:
I don't feel strongly about it...I'm ok leaving it as is.

geoffrey4444 commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -33,52 +33,77 @@ namespace Poisson {
- * \details This system formulates the Poisson equation \f$-\Delta u(x) =
- * f(x)\f$ as the set of coupled first-order PDEs
+ * \details This system formulates the Poisson equation \f$-\Delta_\gamma u(x) =
+ * f(x)\f$ on a background metric \f$\gamma\f$ as the set of coupled first-order
```
\f$\gamma_{ij}\f$

geoffrey4444 commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -33,52 +33,77 @@ namespace Poisson {
+ * second-order (or _primal_) formulation. Furthermore, the linear operator that
+ * represents the DG discretization for this system is not symmetric. This
+ * property further increase the computational cost (see \ref LinearSolverGroup)
+ * and is remedied in the second-order formulation.
```
Will preconditioning be more complicated for this first-order formulation?

nilsvu replied:
No, preconditioning doesn't care about the formulation :) It's an operation on the linear operator and agnostic to how it's built.

geoffrey4444 reviewed (changes requested):
More comments...still need to finish looking through Test_InitializeFluxes.cpp and Test_Equations.cpp in 579b2de and then the bottom 3 commits

nilsvu commented:
@geoffrey4444 I pushed fixups and responded to your comments. Please take a look :)

kidder commented:
@geoffrey4444 please look at the fixups

kidder commented:
@wthrowe can you look over the last two commits of this PR today or tomorrow?  I went over the first 9 commits and have no requested changes.  And Geoffrey has already approved.

wthrowe commented on src/Elliptic/DiscontinuousGalerkin/InitializeFluxes.hpp:
```diff
@@ -30,45 +34,50 @@ namespace Actions {
+      // On exterior (ghost) boundary faces we compute the fluxes from the
+      // data that is being set there to impose boundary conditions. Then, we
+      // compute their normal-dot-fluxes. The flux divergences are sliced from
+      // the volume.
```
I just want to verify that this slicing is the correct behavior, and not a hack like what we have in the evolution code.  If it is not correct please file an issue or comment on the existing boundary condition issue or something.

nilsvu replied:
It‘s the right behavior on Dirichlet boundaries, which is all that‘s currently implemented. I‘ll have to revisit this when we add more general boundary conditions.

wthrowe replied:
OK.  Thanks for confirming.

wthrowe commented on src/Elliptic/Systems/Poisson/Equations.cpp:
```diff
@@ -23,50 +23,23 @@
+                          flux_for_gradient,
+                      const Scalar<DataVector>& field) noexcept {
+  *flux_for_gradient =
+      make_with_value<tnsr::Ij<DataVector, Dim, Frame::Inertial>>(field, 0.);
```
Only do a resize if necessary.

nilsvu replied:
This sets all non-diagonal components to zero. Is there a better way?

wthrowe replied:
I see `std::fill` used in some places, and I can't think of anything better.

wthrowe commented on src/Elliptic/Systems/Poisson/Equations.hpp:
```diff
@@ -47,55 +51,63 @@ class er;
+  static void apply(
+      const gsl::not_null<Scalar<DataVector>*> source_for_field,
+      const Scalar<DataVector>& field) noexcept {
+    *source_for_field = make_with_value<Scalar<DataVector>>(field, 0.);
```
Only resize if necessary.

nilsvu replied:
same here, it zeroes the source

wthrowe replied:
This one can just use `get(...) = 0.` since it's a scalar.  (After resizing if necessary.)

wthrowe commented on src/Elliptic/Systems/Poisson/FirstOrderSystem.hpp:
```diff
@@ -33,50 +34,81 @@ namespace Poisson {
+ * by choosing the fluxes and sources in terms of the system variables
+ * \f$u(x)\f$ and \f$v_i(x)\f$ as
+ *
+ * \f{*align}
```
Rendering error.

wthrowe reviewed (changes requested):
I reviewed the topologically last two commits:  "Formulate elliptic system in terms of fluxes and sources" and "Use Poisson field gradient for auxiliary variable".  (I had no comments on the latter of those two.)  If you wanted me to look at the (author-)temporally last two commits instead, let me know.

kidder commented:
@wthrowe sorry for not being clear, please also look at "Add first-order elliptic operator"

wthrowe commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -0,0 +1,75 @@
+ * can be added to \f$A(u)\f$ in a subsequent action to build the full DG
+ * operator action.
+ *
+ * We generally build the operator $A(u)$ to perform elliptic solver iterations.
```
Missing `\f`s.

wthrowe commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -0,0 +1,75 @@
+ *   - `operator_tag`
+ */
+template <size_t Dim, template <typename> class OperatorTag, typename VarsTag>
+struct FirstOrderOperator {
```
The documentation for this struct does not seem to be describing the correct thing.  It keeps talking about actions and describes an action interface, but this isn't an action.

nilsvu replied:
made some changes to the docs. Lacking a better name I refer to it as an _invokable_ now. I don't know if we have decided on a name for these classes that conform to `mutate_apply`.

nilsvu commented:
@wthrowe @kidder please check the fixups

wthrowe commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -19,22 +19,22 @@ namespace elliptic {
  * independent of the variables \f$u\f$. In a DG setting, boundary contributions
- * can be added to \f$A(u)\f$ in a subsequent action to build the full DG
- * operator action.
+ * can be added to \f$A(u)\f$ in a subsequently to build the full DG operator.
```
Something is wrong with the wording here.

wthrowe commented on src/Elliptic/FirstOrderOperator.hpp:
```diff
@@ -0,0 +1,75 @@
+namespace elliptic {
+
+/*!
+ * \ingroup ActionsGroup
```
Not an action.

nilsvu commented:
@wthrowe squashed corrections for your last two comments into the `fixup FirstOrderOperator`. Are you and @kidder fine with squashing the fixups?

wthrowe commented:
I am fine with this.

kidder commented:
me too

kidder commented:
@wthrowe this should be ready to merge

nilsvu commented:
@geoffrey4444 please re-approve as well when you're happy with this.

wthrowe reviewed (approved):
Waiting for approval to merge by @geoffrey4444 per author request.

nilsvu commented:
Thanks for reviewing @geoffrey4444 @wthrowe @kidder! The story continues in #1671 where I add a generic numerical flux for elliptic systems. Could you be the reviewers for that one as well?
